seapath_ros_driver/src/seapath_ros_driver.cpp

```diff
@@ -12,163 +12,178 @@ namespace seapath
+                nav_pub_->publish(get_navsatfix_message(data, time));
+                odom_pub_->publish(get_odometry_message(data, time));
+                publish_dyn_tf(data, time);  
+                        
+
+                break;
             }
-
-            std::unique_lock<std::mutex> lock(mutex_); // Thread-safe handling
-            process_kmbinary_data(shared_vector_);
-            publish(data_);
-
-            packet_ready_ = false;
-            lock.unlock();
-        }
-        catch (const std::exception &e)
-        {
-            std::cerr << e.what() << '\n';
+            std::this_thread::sleep_for(std::chrono::milliseconds(100));
         }
     }
 
-    void Driver::publish(KMBinaryData data)
+    void Driver::initial_setup()
     {
+        reset_origin();
+        // Start the timer callback loop after the origin is set
+        timer_ = this->create_wall_timer(std::chrono::milliseconds(10), std::bind(&Driver::timer_callback, this));
+    }
 
-        auto odom = get_odometry_message(data);
-        auto diagnostic_array = get_diagnostic_array(data);
-        auto navSatFix = get_navsatfix_message(data);
-        auto KMBinaryData = get_kmbinary_message(data);
-        auto transform = get_transform_message(data);
+    void Driver::timer_callback()
+    {
+       if(socket_.get_data_status()){
+           auto data = socket_.get_kmbinary_data();
+           auto time = this->get_clock()->now();
 
-        diagnosticArray_pub_->publish(diagnostic_array);
+            kmbinary_pub_->publish(get_kmbinary_message(data));
+            diagnostic_pub_->publish(get_diagnostic_array(data, time));
+            if(!data.status_ok()){
+                return;
+            }
+            nav_pub_->publish(get_navsatfix_message(data, time));
+            odom_pub_->publish(get_odometry_message(data, time));
+            publish_dyn_tf(data, time); 
 
-        // Only publish if socket is connected
-        // Could be changed to check pose and orientation error
-        if (socket_connected_) 
-        {
-            odom_pub_->publish(odom);
-            tf_broadcaster_->sendTransform(transform);
-            nav_pub_->publish(navSatFix);
-            kmbinary_pub_->publish(KMBinaryData);
+        
         }
     }
 
-    void Driver::process_kmbinary_data(std::vector<uint8_t> data)
+    void Driver::reset_origin_callback([[maybe_unused]] const std::shared_ptr<std_srvs::srv::Trigger::Request> request,
+                               std::shared_ptr<std_srvs::srv::Trigger::Response> response)
     {
-        data_ = parse_kmbinary_data(data);
+        if(!socket_.socket_connected()){
+            response->success = false;
+            response->message = "Cannot reset origin when socket is disconnected";
+            return;
+        }
+
+        if(timer_){
+        timer_.reset();
+        }
+        std::this_thread::sleep_for(std::chrono::milliseconds(200)); // Wait for timer to finish callback
```

Comment:
This is probably fine, but there might be a cleaner/more robust solution with [std::future](https://cplusplus.com/reference/future/future/)

Reply:
std::future didn't really fit this problem. Fixed with mutex and locks to be more robust